### ISSUE TITLE
Could you help explain in extract_local_patterns Line 60

### ISSUE THREAD

**Issue by GabbySuwichaya** [cavalli1234/AdaLAM#10]
Hi @cavalli1234 @Dawars  ! 

Thanks so much for sharing the code. I love reading your paper, and it proves that there are so many good combinations that can create a state of the art performance. I am learning about it from your paper. 

Here,  I tried to understand how the outlier rejection works from the code. However, I have stumped upon [extract_local_patterns Line 60](https://github.com/cavalli1234/AdaLAM/blob/47de1dd620734b3c42a852d093bbba2285334de5/adalam/core.py#L60). I wonder what are you trying to do ? Could you please help explain?  
```
Line:60 expanded_local_scores = scores[tokp1] + ransidx.type(scores.dtype)
```
I wonder why you have added them together? 
The score and ransidx are variables that are used for a totally different purpose. The score was containing the ratio of the lowest value and the second-lowest value (I guess it is related to Lowe's thresholding). Meanwhile, the ransidx was from Line 51, which are the row of  `fnn_to_seed_local_consistency_map_corr` whose number of neighbors are higher than a specified value of inliers, or specifically, it is...: 
```
**Line:51**    ransidx, tokp1 = torch.where(fnn_to_seed_local_consistency_map_corr)
```

Or is this the way you re-order to be the same as ransidx? Could you please kindly explain....  :)

**Comment by cavalli1234** [comment]
Hi @GabbySuwichaya!

Your intuition is correct, this is indeed the way I sort samples internally to each independent ransac problem. So here is how it works:

As you already mentioned, `scores` holds the descriptor distance ratios, but any scoring in the range [0, 1] could be plugged in. Here the assumption is that correspondences with lower scores are more likely to be correct than others, so we want them to be more likely to be sampled in following RANSACs. The RANSAC implementation in AdaLAM (inspired by PROSAC) samples points deterministically giving priority to the first examples: https://github.com/cavalli1234/AdaLAM/blob/47de1dd620734b3c42a852d093bbba2285334de5/adalam/ransac.py#L89

So we want low scores to be first. However, just sorting the scores would mix up all local RANSAC problems which should be instead solved independently in parallel. What prevents mixing up distinct problems is the `ransidx` which literally identifies each problem with an integer index.

`ransidx` comes from:
https://github.com/cavalli1234/AdaLAM/blob/47de1dd620734b3c42a852d093bbba2285334de5/adalam/core.py#L51

where `fnn_to_seed_local_consistency_map_corr` comes from function `extract_neighborhood_sets` which assigns keypoints to seeds. `fnn_to_seed_local_consistency_map_corr` is a (num_seeds, num_keypoints) boolean matrix which holds the assignment. Keep in mind that each seed with its local neighborhood is associated to a specific local RANSAC problem, so the index of the seed is the index of the distinct RANSAC problem. This is how the `where` function extracts the ransac identifiers indices at the first output. `tokp1` is instead indexed over all keypoints, so it can be used to find the corresponding score of each keypoint in its corresponding ransac problem with `score[tokp1]`.

So as you said, `scores` and `ransidx` have very different and apparently incompatible meanings. However, we want the output to be sorted by `scores` with consecutively equal `ransidx`. Given that `scores` are always smaller or equal to 1 and `ransidx` is integer, the two signals never mix up when summed and allow to actually obtain the desired sorting and apply a deterministic variant of PROSAC sampling.

Let me know if the explanation is clear or any further is needed! And thank you for the interest!

**Comment by GabbySuwichaya** [comment]
@cavalli1234 Thank you so much for the information and the clarification.      

> Keep in mind that each seed with its local neighborhood is associated to a specific local RANSAC problem, so the index of the seed is the index of the distinct RANSAC problem. 

Is this related to preparing the index for which each of the different local neighborhoods will have different RANSAC parameters in the paper?

**Comment by GabbySuwichaya** [comment]
Also, could you please recommend the parameter settings of AdaLAM for the case where the number of keyponts is lower than 2000?  Let's say 512, 1,024, and 2048 keypoints. 

At the moment, I am using the default setting, which I guess is the setting at 8,000 keypoints ...  The best case of your method is 8,000 keypoints (on the Image Matching challenge) is that right? 

However, I would like to give it the best scenarios at the lower number of keypoints as well. If you have separate sets of parameters for the Image Matching challenge and for the general case, please feel free to share both... 

I  would like to use your method at its best.....

**Comment by cavalli1234** [comment]
@GabbySuwichaya 

yes exactly the `ransidx` is the same index as the i in seeds S_i and their neighborhoods N_i in the paper, so all of their respective RANSACs are going to be independent.

For the parameter settings, I generally found these defaults to be quite robust to changing setups. The main issue with few keypoints is that the local statistics are poorer and provide less discriminative information to separate outliers properly. So a significant drop in performance for adalam is expected with decreasing number of keypoints. I do not have specific parameters for working with few keypoints, but I would definitely look into tuning parameters to increase neighborhood sizes (first increase `search_expansion` for increasing with same seeds and more overlap or increase `area_ratio` for increasing with same overlap and less seeds if you want to keep it fast). Larger neighborhoods generally mean less strictly affine patterns, so it might be necessary to also reduce the `min_confidence` to account for that.

**Comment by GabbySuwichaya** [comment]
I see. Thank you so much. 
Your explanation does help me understand a lot. I will close the issue for now. 
When I have more questions I will come back again.
